### PR TITLE
Feature/4 adding functionality for creating company

### LINKED ISSUE #4: Adding functionality for Creating Company
Need to create an endpoint for creating of Company.

### **Create Company**
**Type**: POST
**Route**: HOST/Api/Company
**Parameters**: JSON-object with new Company information.
{
   "Name": ""
}
**Return value**: New Company Id

### **Validation:**

1. Name - can't be empty, max length 255

If Name is empty the system should return response with a status code 400 and the message "'Name' must not be empty."
If the Name length biggest than 255 the system should return response with a status code 400 and the message "The length of 'Name' must be 255 characters or fewer. You entered {current_length} characters."

### PR DESCRIPTION
**Pull request by Kitaety**
Resolve creation of new controller with a POST endpoint. 
Add auto-mapper and fluent-validation

### CODE REVIEW

File: KitMonitor/Server/ActionFilters/BaseValidationFilter.cs
Comment (MisisSweet) [inline KitMonitor/Server/ActionFilters/BaseValidationFilter.cs:1]: Fix format file

File: KitMonitor/Server/AutoMappers/CompanyAutoMapperProfile.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿using AutoMapper;
+using KitMonitor.Server.Models.Database.Entities;
+using KitMonitor.Server.Models.Dto;
+using KitMonitor.Shared.Models.Entities;
+
+namespace KitMonitor.Server.AutoMappers;
+
+public class CompanyAutoMapperProfile: Profile
```
Comment (MisisSweet) [inline KitMonitor/Server/AutoMappers/CompanyAutoMapperProfile.cs]: there is no indentation

File: KitMonitor/Server/Controllers/CompanyController.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+using KitMonitor.Server.Models.Dto;
+using KitMonitor.Server.Services.Interfaces;
+using KitMonitor.Shared.Models.Api.Requests;
+using Microsoft.AspNetCore.Mvc;
+
+namespace KitMonitor.Server.Controllers;
+
+[ApiController]
+[Route("api/[controller]")]
+public class CompanyController: ControllerBase, ICompanyController
```
Comment (MisisSweet) [inline KitMonitor/Server/Controllers/CompanyController.cs]: https://github.com/Kitaety/KitMonitor/pull/5/files#r1197662719

File: KitMonitor/Server/Controllers/Interfaces/ICompanyController.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿using KitMonitor.Shared.Models.Api.Requests;
+
+namespace KitMonitor.Server.Controllers.Interfaces;
+
+public interface ICompanyController: IBaseController<CompanyRequest>
```
Comment (MisisSweet) [inline KitMonitor/Server/Controllers/Interfaces/ICompanyController.cs]: https://github.com/Kitaety/KitMonitor/pull/5/files#r1197662719

File: KitMonitor/Server/Middleware/DatabaseTransactionMiddleware.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+		}
+		catch
+		{
+			await transaction.RollbackAsync();
+			throw;
+		}
+	}
+}
+
+public static class DatabaseTransactionMiddlewareExtensions
```
Comment (MisisSweet) [inline KitMonitor/Server/Middleware/DatabaseTransactionMiddleware.cs]: it is better to put extension classes in a separate place

File: KitMonitor/Server/Middleware/DatabaseTransactionMiddleware.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	{
+		_next = next;
+	}
+
+	public async Task InvokeAsync(HttpContext httpContext, DatabaseContext databaseContext)
+	{
+		var transaction = await databaseContext.Database.BeginTransactionAsync();
+
+		try
+		{
```
Comment (MisisSweet) [inline KitMonitor/Server/Middleware/DatabaseTransactionMiddleware.cs:19]: need loger

File: KitMonitor/Server/Models/Database/Entities/Indicator.cs [hunk 6]
```diff
@@ -1,27 +1,26 @@
...
-        Ph,
-        Tds,
-        WaterFlow
-    }
+	[ForeignKey("IdMonitorSystem")]
+	public MonitorSystem MonitorSystem { get; set; } = null!;
+	public IEnumerable<IndicatorLog> Log { get; set; } = new List<IndicatorLog>();
 }
+
+public enum IndicatorTypeEnum
```
Comment (MisisSweet) [inline KitMonitor/Server/Models/Database/Entities/Indicator.cs:20]: for transfers, you need to add numbers:
       Temperature = 1,
	Ph = 2,
	Tds = 3,
	WaterFlow = 4

File: KitMonitor/Shared/Models/Api/Requests/CompanyRequest.cs [hunk 7]
```diff
@@ -0,0 +1,6 @@
+﻿using KitMonitor.Shared.Models.Entities;
+
+namespace KitMonitor.Shared.Models.Api.Requests;
+
+public class CompanyRequest: BaseRequest<ClientCompany>
```
Comment (MisisSweet) [inline KitMonitor/Shared/Models/Api/Requests/CompanyRequest.cs]: https://github.com/Kitaety/KitMonitor/pull/5/files#r1197662719

File: KitMonitor/Server/Services/Interfaces/ICompanyService.cs [hunk 8]
```diff
@@ -0,0 +1,7 @@
+﻿using KitMonitor.Server.Models.Dto;
+
+namespace KitMonitor.Server.Services.Interfaces;
+
+public interface ICompanyService: IBaseService<CompanyDto>
```
Comment (MisisSweet) [inline KitMonitor/Server/Services/Interfaces/ICompanyService.cs]: https://github.com/Kitaety/KitMonitor/pull/5/files#r1197662719

File: KitMonitor/Server/Validators/Company/CreateCompanyValidator.cs [hunk 9]
```diff
@@ -0,0 +1,12 @@
+﻿using FluentValidation;
+using KitMonitor.Shared.Models.Entities;
+
+namespace KitMonitor.Server.Validators.Company;
+
+public class CreateCompanyValidator : AbstractValidator<ClientCompany>
+{
+    public CreateCompanyValidator()
+    {
+        RuleFor(x => x.Name).NotEmpty().MaximumLength(255);
```
Comment (MisisSweet) [inline KitMonitor/Server/Validators/Company/CreateCompanyValidator.cs]: what is this x?

File: KitMonitor/Server/Services/CompanyService.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
+﻿using KitMonitor.Server.Models.Dto;
+using KitMonitor.Server.Repositories.Interfaces;
+using KitMonitor.Server.Services.Interfaces;
+
+namespace KitMonitor.Server.Services;
+
+public class CompanyService: ICompanyService
```
Comment (MisisSweet) [inline KitMonitor/Server/Services/CompanyService.cs]: https://github.com/Kitaety/KitMonitor/pull/5/files#r1197662719

File: KitMonitor/Server/Repositories/Interfaces/ICompanyRepository.cs [hunk 11]
```diff
@@ -0,0 +1,8 @@
+﻿using KitMonitor.Server.Models.Dto;
+
+namespace KitMonitor.Server.Repositories.Interfaces;
+
+public interface ICompanyRepository: IBaseRepository<CompanyDto>
```
Comment (MisisSweet) [inline KitMonitor/Server/Repositories/Interfaces/ICompanyRepository.cs]: https://github.com/Kitaety/KitMonitor/pull/5/files#r1197662719

File: KitMonitor/Server/Repositories/CompanyRepository.cs [hunk 12]
```diff
@@ -0,0 +1,29 @@
+﻿using AutoMapper;
+using KitMonitor.Server.Models.Database;
+using KitMonitor.Server.Models.Database.Entities;
+using KitMonitor.Server.Models.Dto;
+using KitMonitor.Server.Repositories.Interfaces;
+
+namespace KitMonitor.Server.Repositories;
+
+public class CompanyRepository: ICompanyRepository
```
Comment (MisisSweet) [inline KitMonitor/Server/Repositories/CompanyRepository.cs]: https://github.com/Kitaety/KitMonitor/pull/5/files#r1197662719

File: KitMonitor/Server/AutoMappers/CompanyAutoMapperProfile.cs (hunk 1 above)
Comment (Kitaety) [reply KitMonitor/Server/AutoMappers/CompanyAutoMapperProfile.cs]: fixed

File: KitMonitor/Server/Controllers/CompanyController.cs (hunk 2 above)
Comment (Kitaety) [reply KitMonitor/Server/Controllers/CompanyController.cs]: fixed

File: KitMonitor/Server/Controllers/Interfaces/ICompanyController.cs (hunk 3 above)
Comment (Kitaety) [reply KitMonitor/Server/Controllers/Interfaces/ICompanyController.cs]: fixed

**Review by MisisSweet — APPROVED** [review]

**Review by MisisSweet — APPROVED** [review]

**Review by MisisSweet — APPROVED** [review]
good

**Pull request merged**